### PR TITLE
rose date: added option to convert ISO8601 durations

### LINKED ISSUE #1575: rose date: support duration as an argument
`rose date` should support conversion of an ISO8601 duration from/to e.g. number of seconds (where possible).

See also metomi/isodatetime#49.

### PR DESCRIPTION
**Pull request by oliver-sanders**
Added functionality to convert ISO8601 durations into seconds, minutes or hours
closes #1575 
@matthewrmshin please review
@benfitzpatrick please review

### CODE REVIEW

File: t/rose-date/00-basic.t [hunk 1]
```diff
@@ -300,4 +300,13 @@ file_cmp "$TEST_KEY.err" "$TEST_KEY.err" <<'__ERR__'
...
 __ERR__
 #-------------------------------------------------------------------------------
+# Test ISO8601 duration convertion
+TEST_KEY=$TEST_KEY_BASE-duration
+run_pass "$TEST_KEY" rose date --as-total=s PT1M
+file_cmp "$TEST_KEY.err" "$TEST_KEY.err" </dev/null
+file_cmp "$TEST_KEY.out" "$TEST_KEY.out" <<'__OUT__'
+60.0
+__OUT__
+run_fail "$TEST_KEY.fail" rose date --as-total=y PT1M
```
Comment (matthewrmshin) [inline t/rose-date/00-basic.t]: Should probably have tests for the expected failure.

Do you think we can have a few more test cases?

**Comment by matthewrmshin** [comment]
Test `t/syntax/00-lib.t` is failing with:

```
/home/matt/rose.git/lib/python/rose/opt_parse.py:80:13: E128 continuation line under-indented for visual indent
/home/matt/rose.git/lib/python/rose/opt_parse.py:81:13: E128 continuation line under-indented for visual indent

```

File: t/rose-date/00-basic.t [hunk 2]
```diff
@@ -21,7 +21,7 @@
 #-------------------------------------------------------------------------------
 . $(dirname $0)/test_header
 #-------------------------------------------------------------------------------
-tests 98
+tests 102
```
Comment (matthewrmshin) [inline t/rose-date/00-basic.t]: The test plan needs updating.

**Comment by matthewrmshin** [comment]
Now looks OK.

File: bin/rose-date [hunk 3]
```diff
@@ -75,7 +80,10 @@
 #         Use the value of the ROSE_TASK_CYCLE_TIME environment variable as the
 #         reference time instead of the current time.
 #     --utc, -u
-#         Assume date time in UTC instead of local or other time zones.
+#         Assume date time in UTC instead of local or other time zones. 
+#     --as-total=TIME_FORMAT
```
Comment (benfitzpatrick) [inline bin/rose-date]: Might as well move this to the top of the options (alphabetical order)

File: bin/rose-date [hunk 4]
```diff
@@ -75,7 +80,10 @@
...
+#         Used to express an ISO8601 duration in the specified time format,
+#         hours h, minutes m or seconds s.
```
Comment (benfitzpatrick) [inline bin/rose-date]: I think it is better if we document it upper case - that's what the code stores anyway, and it fits the standard better (PT1H, etc).

File: doc/rose-command.html [hunk 5]
```diff
@@ -938,6 +938,11 @@ <h2 id="rose-date">rose date</h2>
...
 # 2.6 Between 2 specific date times
 rose date 19700101T000000Z 20380119T031407Z
+
+# 3.  Convert ISO8601 duration
+# 3.1 Into the total number of hours (h), minutes(m) or seconds (s)
+#     it represents
+rose date --as-total=s PT1H
 </pre>
 
     <h3>DESCRIPTION</h3>
```
Comment (benfitzpatrick) [inline doc/rose-command.html]: Need --as-total option documentation here. One day we will auto-generate it...

File: lib/python/rose/date.py [hunk 6]
```diff
@@ -371,5 +375,20 @@ def _print_duration(date_time_oper, opts, args):
     print date_time_oper.date_diff_format(opts.print_format, duration, sign)
 
 
+def _convert_duration(date_time_oper, opts, args):
+    """Implement usage 3 of "rose date", convert ISO8601 duration."""
+    time_in_8601 = date_time_oper.duration_parser.parse(args[0])
+    time = time_in_8601.get_seconds()
+    options = {'S': time, 'M': time/60, 'H': time/3600}
+    if opts.duration_print_format.upper() in options:
+        # supplied duration format is valid (upper removes case-sentisitivity)
```
Comment (benfitzpatrick) [inline lib/python/rose/date.py]: Typo

File: t/rose-date/00-basic.t [hunk 7]
```diff
@@ -300,4 +300,34 @@ file_cmp "$TEST_KEY.err" "$TEST_KEY.err" <<'__ERR__'
...
+# Test rose date --as-total runs without error
+TEST_KEY=$TEST_KEY_BASE-as-total
+run_pass "$TEST_KEY" rose date --as-total=s PT1M
+file_cmp "$TEST_KEY.err" "$TEST_KEY.err" </dev/null
+#-------------------------------------------------------------------------------
+# Test rose date --as-total=s P1DT1H1M1S
+TEST_KEY=$TEST_KEY_BASE-as-total-P1DT1H1M1S-s
+run_pass "$TEST_KEY" rose date --as-total=s P1DT1H1M1S
+file_cmp "$TEST_KEY.out" "$TEST_KEY.out" <<'__OUT__'
+90061.0
```
Comment (benfitzpatrick) [inline t/rose-date/00-basic.t]: Should we return this as an integer? Not sure.

Comment (oliver-sanders) [inline t/rose-date/00-basic.t]: Would make sense for converting to seconds certainly, for minutes and hours we would need to floor values though.

File: doc/rose-command.html (hunk 5 above)
Comment (oliver-sanders) [inline doc/rose-command.html]: Have added, also changed h, m, s for H, M, S

File: lib/python/rose/date.py [hunk 8]
```diff
@@ -371,5 +375,21 @@ def _print_duration(date_time_oper, opts, args):
...
+    time_in_8601 = date_time_oper.duration_parser.parse(args[0])
+    time = time_in_8601.get_seconds()
+    options = {'S': time, 'M': time/60, 'H': time/3600}
+    if opts.duration_print_format.upper() in options:
+        # supplied duration format is valid (upper removes
+        # case-sensitivity)
+        print options[opts.duration_print_format.upper()]
+    else:
+        # supplied duration format not valid
+        print 'Invalid date/time format, please use one of h, m, s ' + \
```
Comment (benfitzpatrick) [inline lib/python/rose/date.py]: (needs to be upper case here too)

File: t/rose-date/00-basic.t (hunk 7 above)
Comment (benfitzpatrick) [inline t/rose-date/00-basic.t]: Let's leave this one for the future to handle.

**Comment by oliver-sanders** [comment]
@benfitzpatrick I think everything is sorted on this one.

**Comment by benfitzpatrick** [comment]
OK, great. Just running it through the test battery...

File: doc/rose-command.html [hunk 9]
```diff
@@ -954,6 +959,11 @@ <h2 id="rose-date">rose date</h2>
     <h3>OPTIONS</h3>
 
     <dl>
+      <dt><kbd>--as-total=TIME_FORMAT</kbd><dt>
```
Comment (benfitzpatrick) [inline doc/rose-command.html]: The `t/syntax/01-doc.t` rose test-battery test notices that this should be a `</dt>` - I didn't!

Comment (oliver-sanders) [inline doc/rose-command.html]: Fixed

**Pull request merged**